xarray/core/parallel.py

```diff
@@ -248,8 +292,29 @@ def _wrapper(func, obj, to_array, args, kwargs):
         input_is_array = False
 
     input_chunks = dataset.chunks
+    dataset_indexes = set(dataset.indexes)
+    if template is None:
+        # infer template by providing zero-shaped arrays
+        template = infer_template(func, obj, *args, **kwargs)
+        template_indexes = set(template.indexes)
+        preserved_indexes = template_indexes & dataset_indexes
+        new_indexes = template_indexes - dataset_indexes
+        indexes = {dim: dataset.indexes[dim] for dim in preserved_indexes}
+        indexes.update({k: template.indexes[k] for k in new_indexes})
+        output_chunks = {
+            dim: input_chunks[dim] for dim in template.dims if dim in input_chunks
+        }
+
+    else:
+        # template xarray object has been provided with proper sizes and chunk shapes
```

Comment:
Can we add a type check to verify that `template` is an xarray object? I can imagine users trying something like a tuple of DataArray objects and getting a strange error.